### PR TITLE
Enhance URI errors

### PR DESCRIPTION
**Pull request by chrysn**
Closes: https://github.com/chrysn/aiocoap/pull/364

This should enhance the user experience around entering non-URI addresses:

```
$ aiocoap-client 'example.net/foo'
URL incomplete: Must start with a scheme.
Debugging hint: Most URLs in aiocoap need to be given with a scheme, eg. the 'coap' in 'coap://example.com/path'.
$ aiocoap-client 'https://example.net/foo'
Error: No remote endpoint set for request.
Debugging hint: The message is set up for use with a proxy (because the scheme of 'https://example.net/foo' is not supported), but no proxy was set.
$ aiocoap-client 'coap://[/'
Malformed URL: Invalid IPv6 URL
$ aiocoap-client 'coap://localhost/%ff'
Malformed URL: Percent encoded strings in CoAP URIs need to be UTF-8 encoded
```

(Note that following the hint to set a proxy sends you off into other bug country 🤦)

As bycatch, this cleans up very nonstandard behavior and error behavior of set_request_uri -- a few CoAP violating URIs are now rejected, and all exceptions are consistently MalformedUriError or IncompleteUriError (both of which are ValueError; previously raised AttributeError in some places were just wrong).

@teufelchen1, does this address your issues of #364?

### CODE REVIEW

**Comment by chrysn** [comment]
Just in case you want to review the code changes (what I'm asking is commentary on the errors): The changes to client.py ("hoist error pretty-printing") are best viewed as https://github.com/chrysn/aiocoap/pull/365/commits/75f0b07c43d15a49de85180725785ea90fc7da3e?diff=unified&w=1 with ignore-whitespace on.

File: tests/test_commandline.py [hunk 1]
```diff
@@ -215,6 +215,22 @@ def _test_erroneous(self):
...
 
+        try:
+            # No full URI given
+            subprocess.check_output(
+                AIOCOAP_CLIENT + [self.servernetloc + "/empty"],
+                stderr=subprocess.STDOUT,
+            )
+        except subprocess.CalledProcessError as e:
+            self.assertTrue(
+                "Most URLs in aiocoap need to be given with a scheme"
```
Comment (Teufelchen1) [inline tests/test_commandline.py:226]:
```suggestion
                "URL incomplete: Must start with a scheme."
```
This switches from `extra_help()` to the exception description which is hopefully less brittle - and much closer to `with self.assertRaises(IncompleteUrlError):` which would be the accurate test. (But going through subprocess is the right call here)

**Review by Teufelchen1 — COMMENTED** [review]
This looks much better.
You added a test for `IncompeteUrlError` only but introduced `MalformedUrlError` and `MissingRemoteError` as well. Is that intentional?

File: tests/test_commandline.py (hunk 1 above)
Comment (chrysn) [reply tests/test_commandline.py:226]: Thanks, testing for both now – this should help make sure that I don't lose the extra_help output to some small bug. Of course it's brittle, but then also easy to adjust.

**Comment by chrysn** [comment]
There's no 100% coverage in this project, so it's about as intentional as all the choices about coverage, but here it would be good to do more, so extended. (And as these things go, the more you test, the more corner cases you find to do better).

Thanks for the review!
